### PR TITLE
Salesforce Data Source: Filter Operator: support dynamic where clause.

### PR DESCRIPTION
**Pull request by hidekoji**
# Description

For Salesforce Data Source support dynamic where clause to support select "All" option and relative dates such as Last N years.

# Checklist
Make sure you have performed following items before submitting this pull request.
If not, please describe the reason.  

- [x] Add test cases for this fix/enhancement
- [ ] Pass devtools::check()
- [ ] Pass devtools::test()
- [ ] Test installing from github
- [x] Tested with Exploratory

### CODE REVIEW

File: R/salesforce.R [hunk 1]
```diff
@@ -78,7 +78,12 @@ querySalesforceDataWithQuery <- function(server = NULL, username, password, secu
 #' @param securityToken - (optional) security token to login
 #' @param table - The table you want to get data
 #' @param columns - list of columns that you want to get data from the table.
-querySalesforceDataFromTable <- function(server = NULL, username, password, securityToken = NULL, table = NULL, columns = NULL, guessType = TRUE, conditions = NULL, limit = NULL){
+#' @param dataType - list of column data types for the above columns parameter
+#' @param guessType - flag to tell that dataType should be detected. If FALSE is passed, all the data types become character.
+#' @param conditions - (optional) SOQL where conditions.
+#' @param limit - number of rows to return.
+#' @param logicalOperator (optional) - required if the conditions are set. By default it uses "AND" logical operator to construction SOQL where clause.
```
Comment (hideaki) [inline R/salesforce.R:85]: Check the actual values of "AND", "OR".

File: R/salesforce.R [hunk 2]
```diff
@@ -90,10 +95,58 @@ querySalesforceDataFromTable <- function(server = NULL, username, password, secu
...
+      # Check if the filter condition uses parameter (i.e. detect @{param})
+      hasParameter <- stringr::str_detect(conditions[i], "\\@\\{([^\\}]+)\\}")
+      # make sure to resolve parameter @{} syntax.
+      condition <- glue_exploratory(conditions[i], .transformer=salesforce_glue_transformer, .envir = parent.frame())
+      # make sure to resolve ${} syntax for Salesforce filter.
+      condition <- glue_salesforce(condition)
+      # When the IN (NULL) condition is detected after resolving parameter, remove the condition to support select "ALL" option.
+      # NOTE: IN (@{PARAM}) became IN (NULL) when nothing is selected from UI and Salesforce SOQL does not allow using IS_NULL function,
+      # so remove the "Column IN (NULL)" condition.
+      if (hasParameter && stringr::str_detect(condition, "IN \\(NULL\\)")) {
```
Comment (hideaki) [inline R/salesforce.R:111]: Make it more strict by detecting the string only at the end of the condition.

File: R/salesforce.R [hunk 3]
```diff
@@ -90,10 +95,58 @@ querySalesforceDataFromTable <- function(server = NULL, username, password, secu
...
+        # do not append the condition.
+      } else if (!is_empty(condition)){
```
Comment (hideaki) [inline R/salesforce.R:113]: Distinguish this from rlang::is_empty.

File: R/salesforce.R [hunk 4]
```diff
@@ -90,10 +95,58 @@ querySalesforceDataFromTable <- function(server = NULL, username, password, secu
...
+        whereClause <- stringr::str_c(whereClause, " ", condition)
```
Comment (hideaki) [inline R/salesforce.R:114]: Add condition for i == 1 and other cases.

File: R/salesforce.R [hunk 5]
```diff
@@ -90,10 +95,58 @@ querySalesforceDataFromTable <- function(server = NULL, username, password, secu
...
+    } else if (dataType == "Date") {
+      df[col] <- as.Date(NA)
+    } else if (dataType == "POSIXct") {
+      df[col] <- as.POSIXct(NA)
+    } else {
+      df[col] <- NA
+    }
+  }
+  # Make sure resulting column order is as same as column order selected in UI.
+  df %>% select(columns)
```
Comment (hideaki) [inline R/salesforce.R:151]: Add namespace dplyr::

File: R/system.R [hunk 6]
```diff
@@ -234,6 +234,15 @@ glue_exploratory <- function(text, .transformer, .envir = parent.frame()) {
   ret
 }
 
+# Wrapper around glue::glue() to resolve our ${} notation used for Salesforce Filter.
```
Comment (hideaki) [inline R/system.R:237]: Add comment that this expects @ to be already resolved.

File: R/salesforce.R (hunk 1 above)
Comment (hidekoji) [reply R/salesforce.R]: Updated to lower case since Desktop sends "and" and "or" as lower case.

File: R/salesforce.R (hunk 2 above)
Comment (hidekoji) [reply R/salesforce.R]: Added condition to check if it at the ends of condition.

File: R/salesforce.R (hunk 3 above)
Comment (hidekoji) [reply R/salesforce.R]: Added comment and change it to append namespace (exploratory::)

File: R/salesforce.R (hunk 4 above)
Comment (hidekoji) [reply R/salesforce.R]: Updated condition.

File: R/salesforce.R (hunk 5 above)
Comment (hidekoji) [reply R/salesforce.R]: Added dplyr:: namespace.

File: R/system.R (hunk 6 above)
Comment (hidekoji) [reply R/system.R:237]: Added comment.

**Pull request merged**